src/service/storage-impl.js

```diff
@@ -0,0 +1,391 @@
+          .then(blob => blob ? JSON.parse(atob(blob)) : {})
+          .catch(reason => {
+            log.error(TAG, 'Failed to load store: ', reason);
+            return {};
+          })
+          .then(obj => new Store(obj));
+    }
+    return this.storePromise_;
+  }
+
+  /**
+   * @param {function(!Store)} mutator
+   * @return {!Promise}
+   * @private
+   */
+  saveStore_(mutator) {
+    return this.getStore_()
+        .then(store => {
+          mutator(store);
+          const blob = btoa(JSON.stringify(store.obj));
+          return this.binding_.saveBlob(this.origin_, blob);
+        })
+        .then(this.broadcastReset_.bind(this));
+  }
+
+  /** @private */
+  listenToBroadcasts_() {
+    this.viewer_.onBroadcast(message => {
+      if (message['type'] == 'amp-storage-reset' &&
+              message['origin'] == this.origin_) {
+        log.fine(TAG, 'Received reset message');
+        this.storePromise_ = null;
+      }
+    });
+  }
+
+  /** @private */
+  broadcastReset_() {
+    log.fine(TAG, 'Broadcasted reset message');
+    this.viewer_.broadcast({
+      'type': 'amp-storage-reset',
+      'origin': this.origin_
+    });
+  }
+}
+
+
+/**
+ * The implementation of store logic for get, set and remove.
+ *
+ * The structure of the store is equivalent to the following typedef:
+ * ```
+ * {
+ *   vv: !Array<!{
+ *     k: string,
+ *     v: *,
+ *     t: time
+ *   }>
+ * }
+ * ```
+ *
+ * @private Visible for testing only.
+ */
+export class Store {
+  /**
+   * @param {!JSONObject} obj
+   * @param {number=} opt_maxValues
+   */
+  constructor(obj, opt_maxValues) {
+    /** @const {!JSONObject} */
+    this.obj = obj;
+
+    /** @private @const {number} */
+    this.maxValues_ = opt_maxValues || MAX_VALUES_PER_ORIGIN;
+
+    /** @private @const {!Array<!JSONObject>} */
+    this.values_ = obj['vv'] || [];
+    if (!obj['vv']) {
+      obj['vv'] = this.values_;
+    }
+  }
+
+  /**
+   * @param {string} name
+   * @return {*|undefined}
+   * @private
+   */
+  get(name) {
+    // The structure is [{k: string, v: *, t: time}]
+    for (let i = 0; i < this.values_.length; i++) {
+      if (this.values_[i]['k'] == name) {
+        return this.values_[i]['v'];
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * @param {string} name
+   * @param {*} value
+   * @private
+   */
+  set(name, value) {
+    // The structure is [{k: string, v: *, t: time}]
+    let index = -1;
+    for (let i = 0; i < this.values_.length; i++) {
+      if (this.values_[i]['k'] == name) {
+        index = i;
+        break;
+      }
+    }
+    if (index != -1) {
+      this.values_[index]['v'] = value;
+      this.values_[index]['t'] = timer.now();
+    } else {
+      this.values_.push({'k': name, 'v': value, 't': timer.now()});
+    }
+
+    // Purge old values.
+    if (this.values_.length > this.maxValues_) {
+      let minIndex = 0;
+      for (let i = 1; i < this.values_.length; i++) {
```

Comment:
IE. imagine `0` and `1` are both expired. We'll always splice out `1` with this code, where `0` should be the one to go.